### ISSUE TITLE
ReplyTo always same as From?

### ISSUE THREAD

**Issue by jacargentina** [jstedfast/MailKit#213]
I'm using the IMessageSummary.Envelope.ReplyTo property, and it always returns the same value as IMessageSummary.Envelope.From, even when the message was not using the Reply-To header.

Any advice on this? It is a bug? Or a missuse on my side?

**Comment by jstedfast** [comment]
The easiest way to figure this out is to look at the [protocol log](https://github.com/jstedfast/MailKit/blob/master/FAQ.md#ProtocolLog).

Looking at the code, it does not look possible for the Envelope.ReplyTo property to be set to the same value as the From property unless the server sent the exact same values for both.

**Comment by jacargentina** [comment]
I've enabled logging, and for any given message i get:
````
S: * 83 FETCH (UID 83 MODSEQ (1) INTERNALDATE "14-Dec-2005 04:05:18 +0000" FLAGS (\Seen) ENVELOPE ("Wed, 14 Dec 2005 01:05:18 -0300" "Problema nuestro ? o de HSI?" (("Javier Castro" NIL "javier.alejandro.castro" "gmail.com")) (("Javier Castro" NIL "javier.alejandro.castro" "gmail.com")) (("Javier Castro" NIL "javier.alejandro.castro" "gmail.com")) (("Pablo Lucchetti" NIL "plucchetti" "gmail.com")) NIL NIL NIL "<c711cf0e0512132005v54783f99u@mail.gmail.com>") BODYSTRUCTURE ("TEXT" "PLAIN" ("CHARSET" "ISO-8859-1") NIL NIL "QUOTED-PRINTABLE" 459 14 NIL ("INLINE" NIL) NIL))
````

AFAIK, after ENVELOPE, the third ( ... ) contains the reply-to list: in this case, (and belive me on all cases too), it is always my very same own address (javie...@gmail.com)

Maybe the problem is my understanding of the IMAP protocol? If i send a message without the ReplyTo: heading, the ENVELOPE for that sent message, should/should'nt be empty?

**Comment by jstedfast** [comment]
It sounds like your IMAP server is adding a Reply-To header or filling in the Reply-To field with the From field automatically with the assumption that it would match the From header.

Normally I would say that if you send a message without a Reply-To header, you should get a NIL value for the Reply-To, but your IMAP server apparently does not follow that rule.

**Comment by jacargentina** [comment]
"My" IMAP server is @gmail.com btw.

**Comment by jstedfast** [comment]
FWIW, this is what I get with GMail:

```
C: A00000010 APPEND UnitTests {336}
S: + go ahead
C: From: Example From <from@example.com>
C: To: Example To <to@example.com>
C: Date: Mon, 13 Jul 2015 15:26:08 -0400
C: Subject: 
C: Message-Id: <IB667ICCLVT4.PEVC05CNRU0W2@Jeffrey-Stedfasts-2013-iMac.local>
C: Reply-To: Example Reply-To <reply-to@example.com>
C: MIME-Version: 1.0
C: Content-Type: text/plain; charset=utf-8
C: 
C: This is the message body.
S: A00000010 OK [APPENDUID 56 1] (Success)
C: A00000011 EXAMINE UnitTests (CONDSTORE)
S: * FLAGS (\Answered \Flagged \Draft \Deleted \Seen $Junk $MailFlagBit0 $NotJunk $NotPhishing $Phishing JunkRecorded)
S: * OK [PERMANENTFLAGS ()] Flags permitted.
S: * OK [UIDVALIDITY 56] UIDs valid.
S: * 1 EXISTS
S: * 0 RECENT
S: * OK [UIDNEXT 2] Predicted next UID.
S: * OK [HIGHESTMODSEQ 40284]
S: A00000011 OK [READ-ONLY] UnitTests selected. (Success)
C: A00000012 UID FETCH 1 ENVELOPE
S: * 1 FETCH (UID 1 MODSEQ (40276) ENVELOPE ("Mon, 13 Jul 2015 15:26:08 -0400" NIL (("Example From" NIL "from" "example.com")) (("Example From" NIL "from" "example.com")) (("Example Reply-To" NIL "reply-to" "example.com")) (("Example To" NIL "to" "example.com")) NIL NIL NIL "<IB667ICCLVT4.PEVC05CNRU0W2@Jeffrey-Stedfasts-2013-iMac.local>"))
S: A00000012 OK Success
```

In my case, GMail returned the proper value for the `Reply-To` header in the `ENVELOPE`, but since the `Sender` header did not exist, it should not have re-used the `From` header for the `Sender`, but it did.

In this next example, I do not set a `Reply-To` header and it reproduces exactly what you are seeing (GMail fills in the `Reply-To` value with the `From` header).

```
C: A00000010 APPEND UnitTests {284}
S: + go ahead
C: From: Example From <from@example.com>
C: To: Example To <to@example.com>
C: Date: Mon, 13 Jul 2015 15:28:52 -0400
C: Subject: 
C: Message-Id: <AJQX89DCLVT4.7TZN296JH21Q@Jeffrey-Stedfasts-2013-iMac.local>
C: MIME-Version: 1.0
C: Content-Type: text/plain; charset=utf-8
C: 
C: This is the message body.
S: A00000010 OK [APPENDUID 58 1] (Success)
C: A00000011 EXAMINE UnitTests (CONDSTORE)
S: * FLAGS (\Answered \Flagged \Draft \Deleted \Seen $Junk $MailFlagBit0 $NotJunk $NotPhishing $Phishing JunkRecorded)
S: * OK [PERMANENTFLAGS ()] Flags permitted.
S: * OK [UIDVALIDITY 58] UIDs valid.
S: * 1 EXISTS
S: * 0 RECENT
S: * OK [UIDNEXT 2] Predicted next UID.
S: * OK [HIGHESTMODSEQ 40405]
S: A00000011 OK [READ-ONLY] UnitTests selected. (Success)
C: A00000012 UID FETCH 1 ENVELOPE
S: * 1 FETCH (UID 1 MODSEQ (40397) ENVELOPE ("Mon, 13 Jul 2015 15:28:52 -0400" NIL (("Example From" NIL "from" "example.com")) (("Example From" NIL "from" "example.com")) (("Example From" NIL "from" "example.com")) (("Example To" NIL "to" "example.com")) NIL NIL NIL "<AJQX89DCLVT4.7TZN296JH21Q@Jeffrey-Stedfasts-2013-iMac.local>"))
S: A00000012 OK Success
```

**Comment by jstedfast** [comment]
Ah, here in rfc3501, it states:

```
         If the Sender or Reply-To lines are absent in the [RFC-2822]
         header, or are present but empty, the server sets the
         corresponding member of the envelope to be the same value as
         the from member (the client is not expected to know to do
         this).

            Note: [RFC-2822] requires that all messages have a valid
            From header.  Therefore, the from, sender, and reply-to
            members in the envelope can not be NIL.
```

**Comment by jacargentina** [comment]
So... how can i know the difference between the ENVELOPE reply-to, and the Reply-To: header provided originally by the sender of the message? A simple check like:

    summary.Envelope.From.ToString() != summary.Envelope.ReplyTo.ToString()

should do the magic?

btw, it seems not only gmail, but also cyrus-imapd does that thing.

**Comment by jstedfast** [comment]
It does not look like there is going to be a way to do what you want without requesting the raw headers - you can't assume that just because the From and Reply-To Envelope values match that the Reply-To was auto-populated by the IMAP server since the raw message *could* just have the same values for both.

The reason that Cyrus IMAPd also does it is because it turns out that the IMAP specification says to do things that way (I posted the relevant snippet in my last comment).

What you can do if you really need to check raw header values is this:

```csharp
var summaries = folder.Fetch (0, -1, items, new HashSet<string> () { "Reply-To" });
```

Basically, you can just have MailKit fetch whatever extra raw headers you want and then look at the `summary.Headers` list.

Hope that helps.
